### PR TITLE
Implement match type amendment: extractors follow aliases and singletons

### PR DESCRIPTION
**Pull request by smarter**
This implements the change proposed in
https://github.com/scala/improvement-proposals/pull/84.

The added pos test case presents motivating examples, the added neg test cases demonstrate that errors are correctly reported when cycles are present. The potential for cycle is no worse than with the existing extraction logic as demonstrated by the existing test in `tests/neg/mt-deskolemize.scala`.

### CODE REVIEW

**Comment by smarter** [comment]
I think it'd be useful to get this in 3.5.0-RC1 for testing purposes, we'd then revert it before the release if the sip committee votes against it.

**Comment by Gedochao** [comment]
> I think it'd be useful to get this in 3.5.0-RC1 for testing purposes, we'd then revert it before the release if the sip committee votes against it.

We will not be merging this before SIP committee decision.
Since it won't land in 3.5.0-RC1, we will postpone this until 3.6.0.

**Comment by smarter** [comment]
I'd like to formally appeal that decision. As can be seen in https://github.com/scala/improvement-proposals/pull/84, there is support among the SIP committee members for not unnecessarily delaying this. Delaying makes sense if we believe that it will improve the quality of the final outcome, but we shouldn't stale things for purely procedural reasons. The procedures were put in place to achieve a certain result (build consensus, give enough time for experimentation, ...), so we should always ask ourselves if what we're doing is aligned with the result we want.

**Comment by smarter** [comment]
I've now moved the new logic under an experimental language feature so that it can be merged without SIP approval.

**Review by sjrd — APPROVED** [review]

File: compiler/src/dotty/tools/dotc/core/TypeComparer.scala [hunk 1]
```diff
@@ -3518,20 +3518,80 @@ class MatchReducer(initctx: Context) extends TypeComparer(initctx) {
...
+             */
+            class DropSkolemMap(skolem: SkolemType) extends TypeMap:
+              var refersToSkolem = false
+              def apply(tp: Type): Type =
+                tp match
+                  case `skolem` =>
+                    refersToSkolem = true
+                    tp
+                  case tp: NamedType if betterMatchTypeExtractorsEnabled =>
+                    var savedRefersToSkolem = refersToSkolem
```
[the remark this answers is not in the document]
Comment (smarter) [reply compiler/src/dotty/tools/dotc/core/TypeComparer.scala]: Yes indeed, could still be useful if we want to give more precise error messages but that would be another PR.

**Pull request merged**